### PR TITLE
MAINT: stats.rv_discrete.pmf: should be zero at non-integer argument

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-17146
gh-15932

#### What does this implement/fix?
There was a report in gh-17146 that `rv_discrete.pmf` does not return 0 at non-integer arguments as it used to. This was broken because a line in gh-15932 was out of the intended order. This fixes the order and adds tests.

### CODE REVIEW

File: scipy/stats/tests/test_discrete_basic.py [hunk 1]
```diff
@@ -210,6 +210,17 @@ def check_pmf_cdf(distfn, arg, distname):
     npt.assert_allclose(cdfs - cdfs[0], pmfs_cum - pmfs_cum[0],
                         atol=atol, rtol=rtol)
 
+    # also check that pmf at non-integral args is zero
+    k = np.asarray(index)
+    k_shifted = index[:-1] + np.diff(index)/2
+    npt.assert_equal(distfn.pmf(k_shifted, *arg), 0)
```
Comment (mdhaber) [inline scipy/stats/tests/test_discrete_basic.py:216]:
```suggestion
    # also check that pmf at non-integral k is zero
    k = np.asarray(index)
    k_shifted = k[:-1] + np.diff(k)/2
    npt.assert_equal(distfn.pmf(k_shifted, *arg), 0)
```

**Comment by mdhaber** [comment]
@ev-br, could you take a look here?

Also, you [mentioned](https://github.com/scipy/scipy/issues/8057#issuecomment-1113991534) that you could look into gh-8057 - I'd be happy to review a PR for that!

**Comment by ev-br** [comment]
LGTM! Shall we merge it as is, or is there anything you're adding here still?

**Comment by mdhaber** [comment]
All set, if the last commit looks safe enough to you.

**Comment by ev-br** [comment]
LGTM, let's roll with it. Merged, thanks Matt

**Comment by tylerjereddy** [comment]
This has a defect label and the two commits can apparently be cleanly cherry-picked to `maintenace/1.9.x`, so I'll add a backport label. If you know of a reason not to backport, just remove the label soon-ish.

**Comment by mdhaber** [comment]
Good thing to backport. Thanks!

**Comment by aloctavodia** [comment]
I have been using this "feature" [here](https://preliz.readthedocs.io/en/latest/api_reference.html#module-preliz.distributions.discrete) just as a "trick" for plotting (the dots are evaluated at integers values and the dashed line for floats). I also have being using this "feature " for a project where I approximate discrete variables (actually their pmf) as continuous. 

I understand that the pmf should be zero at non-integer argument and thus this should be the default behaviour. But I will really appreciate an argument to allow results for non-integer values. Is that possible?

**Comment by rkern** [comment]
The behavior that you plot is not intentional, just an accident that the formulae used to compute the integral-input PMFs happen to also be evaluatable on non-integer values. It doesn't necessarily have any particular properties that would make it suitable for approximating PMFs as PDFs or for plots.

**Comment by aloctavodia** [comment]
Thanks for the feedback. In those visualizations the dashed line is used just as visual aid, knowing thatthe dots are the actual values of the pmf. They seem to work reasonable well in practice, at least so far.

There has been some previous interest on evaluating the pmf at non-integers values like this paper https://arxiv.org/abs/2102.04509.

**Comment by rkern** [comment]
What's required in that paper are particular continuous extensions of certain ways to formulate the discrete PMFs. I'm pretty sure that we don't want to take on the burden to guarantee those. It's quite a niche use case, and would likely involve tradeoffs with the main use case.

**Pull request merged**